spec/consensus/reactor/reactor.md

```diff
@@ -0,0 +1,355 @@
+    - If the conditions have changed, a new message is created, effectively superseding the message sent on the previous iteration.
+    - Superseded messages will never be sent through P2P-I again and therefore need not be maintained. (TODO: Argue that this is still valid refinement mapping).
+    - From the point of view of the Gossip layer, Send and TrySend are similar.
+
+- Pub-Sub
+    - Upon certain changes on the Consensus state (detected through internal pub-sub) broadcast messages
+        - These messages are not discarded, even if superseding messages are created, unless there is a reconnection.
+            - Buffer bounding here comes from superseding messages only being generated if the algorithm made some progress, which implies (maybe) some communication with neighbors and therefore that the message was actually sent on the TCP channel?Does not seem really true, since progress means messages coming in, but not necessarily going out. 
+            But if messages are not going out, a disconnection will happen and the buffer will be emptied.
+            - On node reconnection, only the latest (superseding) messages are retransmitted.
+
+Considering the two forms of communication above, Msgs is bounded.
+
+
+
+[GOSSIP-CAST-FORWARD]
+
+- `Receive` function - loop: receive and validate message; invoke PeerState methods to handle contents, updating the state of nodes, which generate superseding messages.
+
+> **Warning**   
+> Are there purely forwarded messages?
+
+
+### Requires from P2P (P2P-I)
+
+[CONCURRENT-CONNECTION]
+
+Ability to connect to multiple nodes concurrently
+
+> P2P provides 1-to-1 interface, while gossip must provide 1-to-many.
+
+[CHURN-DETECTION]
+
+Connection and Disconnection alert
+
+> Gossip needs to keep track of neighbors
+
+[UNAMBIGUOUS-SOURCE-IDENTIFICATION]
+
+Ability to discern sources of messages received
+
+> ~~Needed to identify duplications.~~   
+    - Duplication is handled by having messages in the consensus layer be idempotent   
+> Needed to address sender ins request response/situation
+
+
+[NON-REFUTABILITY]
+
+Non-refutability of messages - Is this a gossip or a consensus requirement?
+
+
+
+[UNICAST]
+- Ability address messages to a single neighbor
+
+#### Current implementations
+
+[CONCURRENT-CONNECTION]
+
+- Inherited from the network stack
+- Driven by PEX and config parameters
+
+[CHURN-DETECTION]
+- `AddPeer`
+- `RemovePeer`
+
+[UNAMBIGUOUS-SOURCE-IDENTIFICATION]
+
+- Node cryptographic IDs
+
+[NON-REFUTABILITY]
+
+- Authentication/Signing
+
+[UNICAST]
+
+- `Send(Envelope)`/`TrySend(Envelope)`
+    - Enqueue and forget. 
+    - Disconnection and reconnection may cause drops. 
+    - Enqueuing may block for a while in `Send`, but not on `TrySend`
+
+
+
+#### Non-requirements
+- Non-duplication
+    - Gossip itself will duplicate messages
+    - Idempotency must be implemented by the application
```

Comment:
I guess here by application you mean consensus?

Reply:
Correct. Rephrased for clarity.